### ISSUE TITLE
80-character limit is outdated

### ISSUE THREAD

**Issue by mars0i** [ocaml/v2.ocaml.org#887]
The style guidelines at https://ocaml.org/learn/tutorials/guidelines.html say that code should fit within 80 characters.  This seems anachronistic in an era when monitors are large or at least high-resolution, and people rarely print out code on paper. 

Of course, if you know that you or someone else is likely to print out your code, then it obviously should fit into 80 characters--but that's not the usual case.  (This follows the general rule of thumb of keeping in mind your readers: humans, and not just compilers, obviously.)

Just my suggestion.  I admit that I am an OCaml newbie, which is why I'm reading style guides.  I can obviously choose not to follow the 80-character rule, but it seemed odd to me given current technology.   Personally, I always carefully choose how wide to make my code depending on what will make it more readable, knowing that longer lines means that someone might have to open a larger editor window to read it.  I am coming from Clojure and Common Lisp, where standard style dictates a lot more indentation, so it's sometimes more difficult to fit into 80 characters.  However, in my limited experience with OCaml, it seems that it's sometimes useful to comment a single line of code with an extended comment, which might go well beyond 80 characters.

**Comment by Chris00** [comment]
The default indentation in OCaml is 2 chars (lisp programs tend to be indented by larger amounts) so the 80 chars limit is not reached that fast.  One may argue that, if 80 chars are reached, the nesting is too deep and the code should be re-factorized to make it easier to understand.  Or maybe, the 80 chars are reached because there are many arguments to a function — such as `printf` — but then splitting the arguments on several lines will usually benefit the reader as well.  I grant you that the exact value is partly a matter of taste but too long lines are definitely hard to read: 70–80 chars per line [is widely regarded as the most characters that you can put on a line and still have it optimal for readers](http://mikeyanderson.com/optimal_characters_per_line).

**Comment by mars0i** [comment]
@Chriss00, thanks very much for responding.  I agree strongly about the point concerning the width of regular text.  I don't feel the same way about code, which I think we read in a different way.   However, if 80 characters is a convention for OCaml, I accept that, and will decide what I want to do with my own code.  Another reason in favor of a narrow width is that although we all have high-resolution monitors, people also sometimes read code on phones.

**Comment by Chris00** [comment]
Not everybody in the community would agree on the 80 chars limit but I think it is a good rule of thumb.  I agree that code is not read in the same way as regular text — if only for syntax highlighting — but if one agrees that the length of the line after indentation should be about 70 chars — because it is difficult to apprehend the whole line at once if longer — and that it is better not to go above 4 levels of nesting (2 chars indentation each on average), one is at 80 chars.  Of course it is not a strict rule but I would argue that breaking it at some places should require making sure that it indeed makes the code more legible.   Let us know how you feel after writing more code in OCaml — for example, showing cases where breaking the "rule" has an obvious benefit.  Also, if you think rewording it would make the intent clearer, feel free to submit a PR!

**Comment by pw374** [comment]
Another point that I'd like to point out: in OCaml, it's naturally rare to need very long lines, whereas in some languages, it is naturally often required to have super long lines. The two reasons that come in mind: naming and repetition.
For instance in some languages, you have things like 
TheNameOfThatClass myVariableOfThatClass = new TheNameOfThatClass(blahblah);
And then it just doesn't make much sense to limit to 80 columns because your hands would be tied in your back!
In OCaml, it's super rare to find this kind of situations.

Philippe Wang
(mail sent from a phone)

**Comment by dbuenzli** [comment]
> This seems anachronistic in an era when monitors are large or at least high-resolution, and people rarely print out code on paper.

As @Chris00 mentions the limit has nothing to do with technology but basic human factors which also happen to exist when you read code. Respecting this limit also helps in improving the readability of your code by imposing you to decompose and name (`let`) larger computations in eye-sized chunks and thus limit the cognitive burden of a fully "inlined" style.

**Comment by dsheets** [comment]
I find that vertical length -- line count -- is more efficient and information dense than horizontal width -- line length. Reading code in a full-screen buffer on an HD monitor isn't a very good use of the right side of the screen as only a few lines will reach that width.

Given the top-to-bottom sequential layout of programming language text, it's beneficial to have a standard maximum width. This lets me fit two 100+ line buffers side-by-side on a 16:10 monitor in portrait orientation with minimal line wrapping. With two portrait monitors side-by-side, I can load 400+ lines of well-formatted code at a good font size to be displayed while I am working. This means that a lot of lookups to interfaces or documentation can be done simply by shifting my gaze.

In short, it's easier to manipulate and display content that has a fixed width but unbounded length. If both width and length are unbounded, it's difficult to display efficiently and without jarring line wrapping.

**Comment by mars0i** [comment]
I understand your point @dsheets.  I sometimes do something similar, but it depends on the code.

What I'm starting to suspect is that it's just a fact that for most people, OCaml is easier to read in short lines, and that larger in-line functions are awkward looking if they can be avoided.  Or maybe it's just what people get used to, but that doesn't make it wrong.  I'm not able to judge this for myself, as I'm still "tuning" my eye for OCaml while adjusting some of my lisp-based habits.  It seems to me that lisps and MLs will naturally have different reading characteristics.  Novices complain about lisp's parentheses, but they probably have certain benefits for human readers, and work well with lisp indentation conventions, just as OCaml's relative lack of parentheses provide other benefits.  

If this is on the right track, then the only thing that I would suggest is that the rationale given for the 80-character guideline be changed.  It's easy to state, but also easy to dispute.  (I still maintain that there is  a place for requirements that code be suitable for printing--but it will be specific to particular courses or collaborations.)

**Comment by dbuenzli** [comment]
> What I'm starting to suspect is that it's just a fact that for most people, OCaml is easier to read in short lines, and that larger in-line functions are awkward looking if they can be avoided.

That's absolutely not specific to OCaml, it's something you'll find [in](http://www.caliban.org/ruby/rubyguide.shtml#linelen) [many](https://www.gnu.org/prep/standards/html_node/Formatting.html#Formatting) [programming](https://www.kernel.org/doc/Documentation/process/coding-style.rst) [guidelines](https://www.python.org/dev/peps/pep-0008/#maximum-line-length), see [here](https://en.wikipedia.org/wiki/Coding_conventions#Coding_conventions_for_languages) for a list to peruse.

> It's easy to state, but also easy to dispute.

So far I didn't see any compelling argument against it.

> I still maintain that there is a place for requirements that code be suitable for printing--but it will be specific to particular courses or collaborations.

Code rarely ends up read by a single person and should always be suitable for printing, if only to be well read on e-readers.

**Comment by mars0i** [comment]
Thanks--I appreciate the links and points you made.  We may not agree on every aspect of this in the end, but readability is important.  

In Clojure, I write >80-column code if and only if it makes my code more intelligible.   There are tradeoffs, but sometimes wider is better.  (This should not seem too odd of a claim given lisp indentation conventions, if you think about it.  (But yes, I break up functions into smaller functions where that seems better.))  For the record, the [Clojure style guide](https://github.com/bbatsov/clojure-style-guide#80-character-limits) also recommends an 80-character limit as a rule of thumb.

If it's less likely that wide code would ever improve readability in OCaml, even on a large high-resolution monitor--as I suspect--then that would be an additional reason to recommend short lines in OCaml, beyond considerations of device width, paper size, etc.  There may be multiple benefits to short lines.

(At this point, for me, at least, I am happiest if I can keep my OCaml code under 50 columns--that seems to make it easier to read for me right now, but as I indicated, OCaml isn't second nature yet.  I like to go beyond 80 only for explanatory comments on individual lines.  I suppose that fully specifying modules of functions would make short lines more difficult, but I try to avoid long inline module specifications--I'm not sure how common that is.)

**Comment by ghost** [comment]
Would be useful to have this commentary linked inline from the style guide.
